internal/client/db/utils.go

```diff
@@ -0,0 +1,201 @@
+// blocks that are in the canonical chain.
+//
+// In the current database schema, this kind of key is only used for
+// lookups into an index, NOT for storing header data or others.
+func newNumberIndexKey[N runtime.Number](num N) (numberIndexKey, error) {
+	if num > N(math.MaxUint32) {
+		return numberIndexKey{}, fmt.Errorf("block number cannot be converted to uint32")
+	}
+	var n uint32 = uint32(num)
+
+	return numberIndexKey{byte(n >> 24), byte((n >> 16) & 0xff), byte((n >> 8) & 0xff), byte(n & 0xff)}, nil
+}
+
+// Convert block id to block lookup key.
+// block lookup key is the DB-key header, block and justification are stored under.
+// looks up lookup key by hash from DB as necessary.
+func blockIDToLookupKey[H runtime.Hash, N runtime.Number](
+	db database.Database[hash.H256], keyLookupCol database.ColumnID, id generic.BlockID,
+) ([]byte, error) {
+	switch id := id.(type) {
+	case generic.BlockIDNumber[N]:
+		key, err := newNumberIndexKey(id.Inner)
+		if err != nil {
+			return nil, err
+		}
+		return db.Get(keyLookupCol, key[:]), nil
+	case generic.BlockIDHash[H]:
+		return db.Get(keyLookupCol, id.Inner.Bytes()), nil
+	default:
+		panic("unsupported generic.BlockIDNumber")
+	}
+}
+
+// Read database column entry for the given block.
+func readDB[H runtime.Hash, N runtime.Number](
+	db database.Database[hash.H256], colIndex, col database.ColumnID, id generic.BlockID,
+) ([]byte, error) {
+	key, err := blockIDToLookupKey[H, N](db, colIndex, id)
+	if err != nil {
+		return nil, err
+	}
+	if key != nil {
+		return db.Get(col, key), nil
+	}
+	return nil, nil
+}
+
+// Read a header from the database.
+func readHeader[H runtime.Hash, N runtime.Number, Header runtime.Header[N, H]](
+	db database.Database[hash.H256], colIndex, col database.ColumnID, id generic.BlockID,
+) (*runtime.Header[N, H], error) {
+	headerBytes, err := readDB[H, N](db, colIndex, col, id)
+	if err != nil {
+		return nil, err
+	}
+	if headerBytes == nil {
+		return nil, nil
+	}
+	var header Header
+	err = scale.Unmarshal(headerBytes, &header)
+	if err != nil {
+		return nil, err
+	}
+	ret := runtime.Header[N, H](header)
+	return &ret, nil
+}
+
+func readMeta[H runtime.Hash, N runtime.Number, Header runtime.Header[N, H]](
+	db database.Database[hash.H256], colHeader database.ColumnID,
+) (meta[H, N], error) {
+	genesisHash, err := readGenesisHash[H](db)
+	if err != nil {
+		return meta[H, N]{}, err
+	}
+	if genesisHash == nil {
+		return meta[H, N]{}, nil
+	}
+
+	var loadMetaBlock = func(desc string, key []byte) (hash H, number N, err error) {
+		id := db.Get(columnMeta, key)
+		if id == nil {
+			return
+		}
+		headerBytes := db.Get(colHeader, id)
+		if headerBytes == nil {
+			return
+		}
+		var header = new(Header)
+		err = scale.Unmarshal(headerBytes, header)
+		if err != nil {
+			return
+		}
+		hash = (*header).Hash()
+		log.Printf("DEBUG: Opened blockchain db, fetched %v = %v (%v)\n", desc, hash, (*header).Number())
```

Comment:
see [this comment](https://github.com/ChainSafe/gossamer/pull/4015#discussion_r1648173688)

I will address this in a future PR, I still want to see everything printed to stdout at the moment.  I'd still want to reference a logger interface rather than concrete logger type.